drivers/sensor/pmw3360/pmw3360.c

```diff
@@ -0,0 +1,938 @@
+
+	ARG_UNUSED(work);
+
+	LOG_DBG("PMW3360 async init step %d", dev_data->async_init_step);
+
+	dev_data->err = async_init_fn[dev_data->async_init_step](dev_data);
+	if (dev_data->err) {
+		LOG_ERR("PMW3360 initailization failed");
+	} else {
+		dev_data->async_init_step++;
+
+		if (dev_data->async_init_step == ASYNC_INIT_STEP_COUNT) {
+			dev_data->ready = true;
+			LOG_INF("PMW3360 initialized");
+		} else {
+			k_delayed_work_submit(&dev_data->init_work,
+					      async_init_delay[dev_data->async_init_step]);
+		}
+	}
+}
+
+static int pmw3360_init_cs(struct pmw3360_data *dev_data)
+{
+	int err;
+
+	dev_data->cs_gpio_dev =
+		device_get_binding(CONFIG_PMW3360_CS_GPIO_DEV_NAME);
+	if (!dev_data->cs_gpio_dev) {
+		LOG_ERR("Cannot get CS GPIO device");
+		return -ENXIO;
+	}
+
+	err = gpio_pin_configure(dev_data->cs_gpio_dev,
+				 CONFIG_PMW3360_CS_GPIO_PIN,
+				 GPIO_DIR_OUT);
+	if (err) {
+		LOG_ERR("Cannot configure CS PIN");
+	}
+
+	return err;
+}
+
+static int pmw3360_init_irq(struct pmw3360_data *dev_data)
+{
+	int err;
+
+	dev_data->irq_gpio_dev =
+		device_get_binding(CONFIG_PMW3360_IRQ_GPIO_DEV_NAME);
+	if (!dev_data->irq_gpio_dev) {
+		LOG_ERR("Cannot get IRQ GPIO device");
+		return -ENXIO;
+	}
+
+	err = gpio_pin_configure(dev_data->irq_gpio_dev,
+				 CONFIG_PMW3360_IRQ_GPIO_PIN,
+				 GPIO_DIR_IN | GPIO_INT | GPIO_PUD_PULL_UP |
+				 GPIO_INT_LEVEL | GPIO_INT_ACTIVE_LOW);
+	if (err) {
+		LOG_ERR("Cannot configure IRQ GPIO");
+		return err;
+	}
+
+	gpio_init_callback(&dev_data->irq_gpio_cb, irq_handler,
+			   BIT(CONFIG_PMW3360_IRQ_GPIO_PIN));
+
+	err = gpio_add_callback(dev_data->irq_gpio_dev, &dev_data->irq_gpio_cb);
+	if (err) {
+		LOG_ERR("Cannot add IRQ GPIO callback");
+	}
+
+	return err;
+}
+
+static int pmw3360_init_spi(struct pmw3360_data *dev_data)
+{
+	dev_data->spi_dev = device_get_binding(DT_SPI_1_NAME);
+	if (!dev_data->spi_dev) {
+		LOG_ERR("Cannot get SPI device");
+		return -ENXIO;
+	}
+
+	return 0;
+}
+
+static int pmw3360_init(struct device *dev)
+{
+	struct pmw3360_data *dev_data = &pmw3360_data;
+	int err;
+
+	ARG_UNUSED(dev);
+
+	err = pmw3360_init_cs(dev_data);
+	if (err) {
+		return err;
+	}
+
+	err = pmw3360_init_irq(dev_data);
+	if (err) {
+		return err;
+	}
+
+	err = pmw3360_init_spi(dev_data);
+	if (err) {
+		return err;
+	}
+
+	k_delayed_work_init(&dev_data->init_work, pmw3360_async_init);
+
+	k_delayed_work_submit(&dev_data->init_work,
+			      async_init_delay[dev_data->async_init_step]);
+
+	return err;
+}
+
+static int pmw3360_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct pmw3360_data *dev_data = &pmw3360_data;
+	u8_t data[PMW3360_BURST_SIZE];
+
+	ARG_UNUSED(dev);
+
+	if (unlikely(!dev_data->ready)) {
+		LOG_INF("Device is not initialized yet");
+		return -EBUSY;
+	}
+
+	int err = motion_burst_read(dev_data, data, sizeof(data));
+
+	if (!err) {
+		dev_data->x =  sys_get_le16(&data[4]);
+		dev_data->y = -sys_get_le16(&data[2]);
```

Comment:
These offsets deserve labels defined for them.
`[2]` is actually `Delta_X_L` and `[4]` is `Delta_Y_L`, so the swapping between `x` and `y` as seen by the sensor and the driver is quite obscured here.
Actually, why is this swapping done here, and the inversion of the `y` value as well? Isn't it a job for the application utilizing the sensor?

Reply:
So long this code was part of the application but you are right. I don't think we should leave this to app. I will add a config option to set the orientation of the sensor.

Reply:
done